AppCenter/AppCenter/Internals/Util/MSEncrypter.m

```diff
@@ -117,52 +65,148 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
+  NSData *cipherText;
+  if (metadata) {
+
+    // Extract key from metadata.
+    NSString *keyTag = [metadata componentsSeparatedByString:kMSEncryptionMetadataInternalSeparator][0];
+
+    // Metadata, separator, and initialization vector.
+    size_t cipherTextPrefixLength = metadataLocation + 1 + kCCBlockSizeAES128;
+    NSRange cipherTextRange = NSMakeRange(cipherTextPrefixLength, [data length] - cipherTextPrefixLength);
+    NSRange ivRange = NSMakeRange(metadataLocation + 1, kCCBlockSizeAES128);
+    initializationVector = [data subdataWithRange:ivRange];
+    cipherText = [data subdataWithRange:cipherTextRange];
+    key = [self getKeyWithKeyTag:keyTag];
   } else {
-    result = [NSData dataWithBytes:cipherBuffer length:numBytesDecrypted];
-    if (!result) {
-      MSLogWarning([MSAppCenter logTag], @"Could not create NSData object from decrypted bytes.");
+
+    // If there is no metadata, this is old data, so use the old key and an empty initialization vector.
+    key = [self getKeyWithKeyTag:kMSEncryptionKeyTagOriginal];
+    cipherText = data;
+  }
+  return [MSEncrypter performCryptoOperation:kCCDecrypt input:cipherText initializationVector:initializationVector key:key];
+}
+
++ (NSString *)getCurrentKeyTag {
+  @synchronized(classLock) {
+    NSString *keyMetadata = [MS_USER_DEFAULTS objectForKey:kMSEncryptionKeyMetadataKey];
+    if (!keyMetadata) {
+      [self rotateToNewKeyTag:kMSEncryptionKeyTagAlternate];
+      return kMSEncryptionKeyTagAlternate;
+    }
+
+    // Format is {keyTag}/{expiration as iso}.
+    NSArray *keyMetadataComponents = [keyMetadata componentsSeparatedByString:kMSEncryptionMetadataInternalSeparator];
+    NSString *keyTag = keyMetadataComponents[0];
+    NSString *expirationIso = keyMetadataComponents[1];
+    NSDate *expiration = [MSUtility dateFromISO8601:expirationIso];
+    BOOL isNotExpired = [[expiration laterDate:[NSDate date]] isEqualToDate:expiration];
+    if (isNotExpired) {
+      return keyTag;
+    }
+
+    // Key is expired and must be rotated.
+    if ([keyTag isEqualToString:kMSEncryptionKeyTagOriginal]) {
+      keyTag = kMSEncryptionKeyTagAlternate;
+    } else {
+      keyTag = kMSEncryptionKeyTagOriginal;
     }
+    [self rotateToNewKeyTag:keyTag];
+    return keyTag;
   }
-  free(cipherBuffer);
-  return result;
 }
 
-+ (void)deleteKeyWithTag:(NSString *)keyTag {
-  [MSKeychainUtil deleteStringForKey:keyTag];
++ (void)rotateToNewKeyTag:(NSString *)newKeyTag {
+  NSDate *expiration = [[NSDate date] dateByAddingTimeInterval:kMSEncryptionKeyLifetimeInSeconds];
+  NSString *expirationIso = [MSUtility dateToISO8601:expiration];
+
+  // Format is {keyTag}/{expiration as iso}.
+  NSString *keyMetadata = [@[ newKeyTag, expirationIso ] componentsJoinedByString:kMSEncryptionMetadataInternalSeparator];
+  [MS_USER_DEFAULTS setObject:keyMetadata forKey:kMSEncryptionKeyMetadataKey];
+}
+
+- (NSData *)getKeyWithKeyTag:(NSString *)keyTag {
+  NSData *key;
+  @synchronized(self) {
```

Comment:
Would it be possible to synchronized the whole block on self at this point. It means it will lock on classLock but that shouldn't introduce a deadlock since classLock doesn't lock on self (it wouldn't make sense). That would simplify the code.

Reply:
We want to keep as much code as possible outside of synchronized blocks for performance reasons though, no? I'd prefer to keep the (only slightly) less readable version if it means a small improvement in performance

Reply:
We could but that would really reduce performance. The way its written means that class lock will almost never be used. I don't think we should change that unless we have to. We can discuss offline if you aren't convinced

Reply:
I also commented synchronization in the method (sadly no answers for the comment).
We need to think about simplifying this. Sounds like it is a trade-off of simplifying/readability vs performance. Can we talk about this one?

Reply:
Removed the lock here altogether. Resolving.